crates/core/src/db/datastore/traits.rs

```diff
@@ -322,6 +322,26 @@ pub struct Program {
     pub bytes: Box<[u8]>,
 }
 
+impl Program {
+    /// Create a [`Program`] from its raw bytes.
+    ///
+    /// This computes the hash over `bytes`, so prefer constructing [`Program`]
+    /// directly if the hash is already known.
+    pub fn from_bytes(bytes: impl Into<Box<[u8]>>) -> Self {
+        let bytes = bytes.into();
+        let hash = hash_bytes(&bytes);
+        Self { hash, bytes }
+    }
+
+    /// Create a [`Program`] with no bytes.
+    pub fn empty() -> Self {
+        Self {
+            hash: Hash::ZERO,
```

Comment:
Not sure if this is used anywhere that would matter, but couldn't hurt to do this:

```suggestion
            hash: hash_bytes([].into()),
```